vg/vgtex/canvas.go

```diff
@@ -0,0 +1,277 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package vgtex provides a vg.Canvas implementation for LaTeX, targeted at the TikZ/PGF LaTeX package:
+//  https://sourceforge.net/projects/pgf
+//
+// vgtex generates PGF instructions that will be interpreted and rendered by LaTeX.
+// vgtex allows to put any valid LaTeX notation inside plot's strings.
+package vgtex
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"image/color"
+	"io"
+	"math"
+	"strings"
+
+	"github.com/gonum/plot/vg"
+)
+
+const degPerRadian = 180 / math.Pi
+
+// Canvas implements the vg.Canvas interface, translating gonum/plot
+// drawing primitives into PGF ones.
+type Canvas struct {
+	buf   *bytes.Buffer
+	w, h  vg.Length
+	dpi   int // dpi is the number of dots per inch for this canvas
+	stack []context
+}
+
+type context struct {
+	color      color.Color
+	dashArray  []vg.Length
+	dashOffset vg.Length
+	linew      vg.Length
+}
+
+// New returns a new LaTeX canvas.
+func New(w, h vg.Length) *Canvas {
+	c := &Canvas{
+		buf: new(bytes.Buffer),
+		w:   w,
+		h:   h,
+		dpi: 72,
+	}
+	c.wtex(`%%%% gonum/plot created for LaTeX/pgf`)
+	c.wtex(`%%%% you need to add:`)
+	c.wtex(`%%%%   \usepackage{pgf}`)
+	c.wtex(`%%%% to your LaTeX document`)
+	c.wtex("")
+	c.wtex(`\begin{pgfpicture}`)
+	c.stack = make([]context, 1)
+	vg.Initialize(c)
+	return c
+}
+
+func (c *Canvas) ctx() *context {
+	return &c.stack[len(c.stack)-1]
+}
+
+// Size returns the width and height of the canvas.
+func (c *Canvas) Size() (w, h vg.Length) {
+	return c.w, c.h
+}
+
+// SetLineWidth sets the width of stroked paths.
+// If the width is not positive then stroked lines
+// are not drawn.
+//
+// The initial line width is 1 point.
+func (c *Canvas) SetLineWidth(w vg.Length) {
+	c.ctx().linew = w
+}
+
+// SetLineDash sets the dash pattern for lines.
+// The pattern slice specifies the lengths of
+// alternating dashes and gaps, and the offset
+// specifies the distance into the dash pattern
+// to start the dash.
+//
+// The initial dash pattern is a solid line.
+func (c *Canvas) SetLineDash(pattern []vg.Length, offset vg.Length) {
+	c.ctx().dashArray = pattern
+	c.ctx().dashOffset = offset
+}
+
+// SetColor sets the current drawing color.
+// Note that fill color and stroke color are
+// the same, so if you want different fill
+// and stroke colors then you must set a color,
+// draw fills, set a new color and then draw lines.
+//
+// The initial color is black.  If SetColor is
+// called with a nil color then black is used.
+func (c *Canvas) SetColor(clr color.Color) {
+	c.ctx().color = clr
+}
+
+// Rotate applies a rotation transform to the
+// context.  The parameter is specified in
+// radians.
+func (c *Canvas) Rotate(rad float64) {
+	c.wtex(`\pgftransformrotate{%g}`, rad*degPerRadian)
+}
+
+// Translate applies a translational transform
+// to the context.
+func (c *Canvas) Translate(x, y vg.Length) {
+	c.wtex(
```

Comment:
I would've thought that gofmt would cuddle the first argument up onto the line with the start of the call. Is this file gofmted?